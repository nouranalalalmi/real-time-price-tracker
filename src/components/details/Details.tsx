import { COINS } from '@/constants';

import { Chart } from './chart/Chart';
import { Overview } from './overview/Overview';

// This function runs at build time and generates the list of all possible paths for dynamic routes
export async function getStaticPaths() {
  const paths = COINS.map(coin => ({
    params: { id: coin.id.toString() },
  }));

  return { paths, fallback: false };
}

// This function runs for each path returned by getStaticPaths and fetches the data for that page
export async function getStaticProps({
  params,
}: {
  params: {
    id: string;
  };
}) {
  const coin = COINS.find(coin => coin.id.toString() === params.id);

  if (!coin) {
    return {
      notFound: true,
    };
  }

  return {
    props: {
      coin,
    },
  };
}

export const Details = () => (
  <>
    <div className="flex w-full flex-col justify-between gap-6 md:flex-row">
      <Chart />
      <div className="w-full md:w-1/4">
        <Overview />
      </div>
    </div>
    <p className="p-2 text-xs font-semibold text-yellow-600">- Data is refetched every 3 minutes</p>
  </>
);
