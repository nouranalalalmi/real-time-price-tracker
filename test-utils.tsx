// test-utils.js
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { RenderOptions, render as rtlRender } from '@testing-library/react';
import { ReactElement } from 'react';

const queryClient = new QueryClient();

function render(ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) {
  return rtlRender(<QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>, options);
}

export * from '@testing-library/react';

export { render };
