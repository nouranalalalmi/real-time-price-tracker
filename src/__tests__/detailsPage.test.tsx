import Details from '@/app/[id]/page';
import { render, screen } from '../../test-utils';
import mockRouter from 'next-router-mock';

describe('Details page', () => {
  const id = 'bitcoin';
  beforeAll(() => {
    mockRouter.push({
      pathname: '/:id',
      query: {
        id,
      },
    });
  });

  it('matches snapshot', () => {
    const { container } = render(<Details />);
    expect(screen.getByText(id)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
