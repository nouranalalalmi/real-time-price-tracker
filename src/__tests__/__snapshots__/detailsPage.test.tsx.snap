// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details page matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-y-2"
  >
    <a
      class="w-fit text-sm font-semibold text-blue-500 hover:text-blue-700"
      href="/"
    >
      ‚Üê back to live ticker
    </a>
    <div
      class="flex w-full flex-col justify-between gap-6 md:flex-row"
    >
      <div
        class="w-full"
      >
        <h1
          class="pb-2 text-3xl font-semibold"
        >
          History
        </h1>
        <div
          class="rounded-lg border bg-white p-2 shadow-sm"
        >
          <div
            class="flex justify-end gap-2"
          >
            <button
              class="border-current/30 rounded-lg border px-2 hover:bg-blue-50 bg-blue-100"
              data-testid="range-1d"
            >
              1d
            </button>
            <button
              class="border-current/30 rounded-lg border px-2 hover:bg-blue-50 false"
              data-testid="range-1w"
            >
              1w
            </button>
            <button
              class="border-current/30 rounded-lg border px-2 hover:bg-blue-50 false"
              data-testid="range-1m"
            >
              1m
            </button>
            <button
              class="border-current/30 rounded-lg border px-2 hover:bg-blue-50 false"
              data-testid="range-1y"
            >
              1y
            </button>
          </div>
          <canvas
            height="150"
            role="img"
            width="300"
          />
        </div>
      </div>
      <div
        class="w-full md:w-1/4"
      >
        <div
          class="flex w-full flex-col"
        >
          <h1
            class="pb-2 text-3xl font-semibold"
          >
            Overview
          </h1>
          <div
            class="flex flex-col gap-y-4 rounded-lg border bg-white p-4 shadow-sm"
          >
            <div
              class="flex items-center gap-2 border-b"
            >
              <h2
                class="text-2xl capitalize"
              >
                bitcoin
              </h2>
              <img
                alt="asset-logo"
                data-nimg="1"
                decoding="async"
                height="25"
                loading="lazy"
                src="/_next/image?url=%2Fimages%2Fbtc.png&w=64&q=75"
                srcset="/_next/image?url=%2Fimages%2Fbtc.png&w=32&q=75 1x, /_next/image?url=%2Fimages%2Fbtc.png&w=64&q=75 2x"
                style="color: transparent;"
                width="25"
              />
            </div>
            <div
              class="size-4 animate-spin rounded-full border-y-2 border-pink-500"
              data-testid="spinner"
            />
          </div>
        </div>
      </div>
    </div>
    <p
      class="p-2 text-xs font-semibold text-yellow-600"
    >
      - Data is refetched every 3 minutes
    </p>
  </div>
</div>
`;
